core/modules/replica/Controller.h

```diff
@@ -0,0 +1,687 @@
+     * and memory management of instances created otherwise (as values or via
+     * low-level pointers).
+     *
+     * @param serviceProvider - for configuration, other services
+     */
+    static pointer create(ServiceProvider::pointer const& serviceProvider);
+
+    // Default construction and copy semantics are prohibited
+
+    Controller() = delete;
+    Controller(Controller const&) = delete;
+    Controller& operator=(Controller const&) = delete;
+
+    /// Destructor
+    ~Controller() = default;
+
+    /// Return the unique identity of the instance
+    ControllerIdentity const& identity() const { return _identity; }
+
+    /// The start time of the instance (milliseconds since UNIX Epoch)
+    uint64_t startTime() const { return _startTime; }
+
+    /// Return the Service Provider used by the server
+    ServiceProvider::pointer const& serviceProvider() { return _serviceProvider; }
+
+    /// @return reference to the I/O service for ASYNC requests
+    boost::asio::io_service& io_service() { return _io_service; }
+
+    /**
+     * Run the Controller in a dedicated thread unless it's already running.
+     * It's safe to call this method multiple times from any thread.
+     */
+    void run();
+
+    /**
+     * Check if the service is running.
+     *
+     * @return true if the service is running.
+     */
+    bool isRunning() const;
+
+    /**
+     * Stop the server. This method will guarantee that all outstanding
+     * opeations will finish and not aborted.
+     *
+     * This operation will also result in stopping the internal thread
+     * in which the server is being run.
+     */
+    void stop();
+
+    /**
+     * Join with a thread in which the service is being run (if any).
+     * If the service was not started or if it's stopped the the method
+     * will return immediattely.
+     *
+     * This method is meant to be used for integration of the service into
+     * a larger multi-threaded application which may require a proper
+     * synchronization between threads.
+     */
+    void join();
+
+    /**
+     * Initiate a new replication request.
+     *
+     * The method will throw exception std::invalid_argument if the worker
+     * names are equal.
+     *
+     * @param workerName       - the name of a worker node from which to copy the chunk
+     * @param sourceWorkerName - the name of a worker node where the replica will be created
+     * @param database         - database name
+     * @param chunk            - the chunk number
+     * @param onFinish         - an optional callback function to be called upon the completion of the request
+     * @param priority         - a priority level of the request
+     * @param keepTracking     - keep tracking the request before it finishes or fails
+     * @param allowDuplicate   - follow a previously made request if the current one duplicates it
+     * @param jobId            - an optional identifier of a job issed the request
```

Comment:
should all the parameters with default values be labeled "optional" in this documentation?

Reply:
(I'll only put the question here, but note that it applies for all the functions in this interface)

Reply:
Fixed